### ISSUE TITLE
`crm cluster stop` must verify and stop sbd.service

### ISSUE THREAD

**Issue by zzhou1** [ClusterLabs/crmsh#872]
There are some situations sbd.service is left running alone while pacemaker and corosync already shutdown.  

For example, when the reboot time of a virtual machine node is too fast, without a proper SBD_DELAY_START value, pacemaker detects "We were allegedly just fenced" and do self "Shutting down Pacemaker". 

In that case, currently,`crm cluster stop` does nothing about sbd.service. It directly returns the following information as long as corosyn.service is inactive, even if sbd.service is still running behind. 

"INFO: Cluster services already stopped"

That says, `crm cluster stop` need proactively check the states for both corosync.service and sbd.service, and reinforce `systemctl stop corosync.service`.

**Comment by wenningerk** [comment]
@zzhou1
Can you be a bit more specific on those cases? Is that somehow intended or rather a situation we should try to have tackled by the cluster-services?

**Comment by zzhou1** [comment]
> @zzhou1
> Can you be a bit more specific on those cases? Is that somehow intended or rather a situation we should try to have tackled by the cluster-services?

**Comment by gao-yan** [comment]
> There are some situations sbd.service is left running alone while pacemaker and corosync already shutdown.
> 
> For example, when the reboot time of a virtual machine node is too fast, without a proper SBD_DELAY_START value, pacemaker detects "We were allegedly just fenced" and do self "Shutting down Pacemaker".

The shutdown is requested by pacemaker which is supposed to shutdown the whole cluster stack at least with recent versions.

> 
> In that case, currently,`crm cluster stop` do nothing about sbd.service, and still leave sbd.service running with the following information:
> 
> "INFO: Cluster services already stopped"
> 
> That says, `crm cluster stop` must proactively stop corosync.service to take down sbd.service.

Anyway, I think `crm cluster stop` does `systemctl stop corosync.service` which takes down sbd.service as well since sb.service is  defined with:
```
PartOf=corosync.service
```
And since sbd.service is also defined with:
```
RefuseManualStop=true
RefuseManualStart=true
```
, it's not supposed to be directly operated by `systemctl start/stop sbd.service`. They should be done together with corosync.service.

**Comment by zzhou1** [comment]
> The shutdown is requested by pacemaker which is supposed to shutdown the whole cluster stack at least with recent versions.
> 

It does take down corosync.service, but not sbd.service. How does pacemaker shutdown procedure talks to corosync ? Sounds like to me, there is no "systemctl stop corosync" during this pacemaker shutdown phase. I'm not familiar with this part at the moment.

Yeah, there is another topic outside of crmsh. 

> > In that case, currently,`crm cluster stop` do nothing about sbd.service, and still leave sbd.service running with the following information:
> > "INFO: Cluster services already stopped"
> > That says, `crm cluster stop` must proactively stop corosync.service to take down sbd.service.

I should rephrase the improvement to be,

`crm cluster stop` need proactively check the states for both corosync.service and sbd.service, and reinforce `systemctl stop corosync.service`. Currently, it just returns if corosyn.service is inactive.

**Comment by wenningerk** [comment]
My experience with crmsh is quite dated (last time I really used it is about 6 years back) so I don't know what's currently happening and most probable don't remember the details what happened 6 years back.
Anyway - yes sbd is controlled by systemd as partOf corosync.
When pacemaker is shutting down this is monitored by sbd - if present of course - and sbd decides if it was a graceful shutdown where the resources are down (actually this isn't checked - sbd believes pacemaker if it tells so - and the admin has to take care of a proper pacemaker configuration or if the node is in maintenance-mode the admin has to know what he is doing).
As already mentioned corosync and with it sbd keep running.
Especially for sbd (with watchdog-fencing) the way pcs is doing a cluster shutdown is interesting: first all pacemaker instances are being shutdown with corosync staying as source for quorum. Just when all pacemaker instances are down - which can take a while as the services in the cluster have to be brought down - it shuts down the corosync-instances - usually so quick that the nodes stay up even with watchdog-fencing sbd configured.
If there are cases where shutting down corosync doesn't as well bring sbd down this is definitely unintentional and should be fixed in corosync, sbd & the way they are working together. I anyway still have to touch that part as using the recently reintroduced shutdown callback from corosync should be used by sbd (pacemaker is already doing that).

**Comment by zzhou1** [comment]
Hi Klaus,

> Anyway - yes sbd is controlled by systemd as partOf corosync.

understand

> When pacemaker is shutting down this is monitored by sbd - if present of course - and sbd decides if it was a graceful shutdown where the resources are down (actually this isn't checked - sbd believes pacemaker if it tells so - and the admin has to take care of a proper pacemaker configuration or if the node is in maintenance-mode the admin has to know what he is doing).
> As already mentioned corosync and with it sbd keep running.

I try to understand the last sentence better, but not get the point very well. Can you rephrase if it is important ;)

> Especially for sbd (with watchdog-fencing) the way pcs is doing a cluster shutdown is interesting: first all pacemaker instances are being shutdown with corosync staying as source for quorum. Just when all pacemaker instances are down - which can take a while as the services in the cluster have to be brought down - it shuts down the corosync-instances - usually so quick that the nodes stay up even with watchdog-fencing sbd configured.

Understand this part.

> If there are cases where shutting down corosync doesn't as well bring sbd down this is definitely unintentional and should be fixed in corosync, sbd & the way they are working together.

Then, there is a bug somewhere indeed :) . BTW, do you need I provide more information to create an issue in sbd project?

> I anyway still have to touch that part as using the recently reintroduced shutdown callback from corosync sounds like a good idea and I haven't gotten to introduce it to sbd (pacemaker is already doing that) yet.

On-and-off, I get to know you are working on something in this area, well, not know the whole picture. Maybe, I need find more time to get to know it better this time.

Today, I get to aware and change SBD_SYNC_RESOURCE_STARTUP to yes. The overall behavior is changed too, sbd do self reboot now, though seems an unexpected from the end user experience perspective. It prints the following messages right before self-fence :

Sep 14 23:16:04 node2 pacemaker-controld[1556]:  crit: We were allegedly just fenced by node1 for node1!
Sep 14 23:16:04 node2 pacemakerd[1548]:  warning: Shutting cluster down because pacemaker-controld[1556] had fatal failure
Sep 14 23:16:04 node2 pacemakerd[1548]:  notice: Shutting down Pacemaker
...
Sep 14 23:16:04 node2 pacemakerd[1548]:  notice: Shutdown complete
Sep 14 23:16:04 node2 pacemakerd[1548]:  notice: Shutting down and staying down after fatal error
Sep 14 23:16:04 node2 corosync[1523]:   [CFG   ] Node 2 was shut down by sysadmin
...
Sep 14 23:16:04 node2 corosync[1523]:   [MAIN  ] Corosync Cluster Engine exiting normally
Sep 14 23:16:04 node2 sbd[1517]:  warning: inquisitor_child: Cluster-Servant has exited (connection lost)
Sep 14 23:16:04 node2 sbd[1517]:  warning: cleanup_servant_by_pid: Servant for cluster (pid: 1521) has terminated
...
Sep 14 23:16:04 node2 sbd[1773]:    cluster:   notice: servant_cluster: Monitoring unknown cluster health
...
Sep 14 23:16:07 node2 sbd[1517]:  warning: inquisitor_child: Servant pcmk is outdated (age: 4)
Sep 14 23:16:07 node2 sbd[1517]:  warning: inquisitor_child: Latency: No liveness for 4s exceeds watchdog warning timeout of 3s (healthy servants: 0)

**Comment by zzhou1** [comment]
A simple reproducer,
```
crm_attribute -t status -n shutdown -v "" 
systemctl status sbd
```

**Comment by liangxin1300** [comment]
In https://github.com/ClusterLabs/crmsh/pull/797/files#diff-528bd5175d453d9de7f004053ff6de322ff5dc282a7fe72a3d51ec2a0073e467R167-R168, we fixed this issue
